view/statistics-flow-certificates.js

```diff
@@ -0,0 +1,322 @@
+  , selectPeriodMode  = require('./components/filter-bar/select-period-mode')
+  , itemsPerPage      = require('../conf/objects-list-items-per-page')
+  , isNaturalNumber   = require('es5-ext/number/is-natural')
+  , serviceQuery      = require('../apps-common/query-conf/service')
+  , certificateQuery  = require('../apps-common/query-conf/certificate');
+
+exports._parent        = require('./statistics-flow');
+exports._customFilters = Function.prototype;
+
+exports['flow-nav']                = { class: { 'submitted-menu-item-active': true } };
+exports['flow-by-certificate-nav'] = { class: { 'pills-nav-active': true } };
+
+var queryServer = memoize(function (query) {
+	return getData('/get-flow-data/', query);
+}, {
+	normalizer: function (args) { return JSON.stringify(args[0]); },
+	max: 1000
+});
+
+var serviceToCertLegacyMatch = { '': [] };
+
+var getServiceName = function (ServiceType) {
+	return uncapitalize.call(
+		ServiceType.__id__.slice('BusinessProcess'.length)
+	);
+};
+
+db.BusinessProcess.extensions.forEach(function (ServiceType) {
+	ServiceType.prototype.certificates.map.forEach(function (certificate) {
+		if (!serviceToCertLegacyMatch[getServiceName(ServiceType)]) {
+			serviceToCertLegacyMatch[getServiceName(ServiceType)] = [];
+		}
+		serviceToCertLegacyMatch[getServiceName(ServiceType)].push('certificate-' + certificate.key);
+		serviceToCertLegacyMatch[''].push('certificate-' + certificate.key);
+	});
+});
+
+var buildResultRow = function (rowData) {
+	return assign({
+		submitted: 0,
+		pending: 0,
+		pickup: 0,
+		withdrawn: 0,
+		rejected: 0,
+		sentBack: 0
+	}, rowData || {});
+};
+
+var accumulateResultRows = function (rows) {
+	var result = buildResultRow(rows[0]);
+	rows.slice(1).forEach(function (row) {
+		Object.keys(row).forEach(function (propertyName) {
+			result[propertyName] += row[propertyName];
+		});
+	});
+
+	return result;
+};
+
+var copyDbDate = function (date) {
+	return new db.Date(date.getUTCFullYear(), date.getUTCMonth(), date.getUTCDate());
+};
+
+var incrementDateByTimeUnit = function (date, mode) {
+	switch (mode) {
+	case 'weekly':
+		date.setUTCDate(date.getUTCDate() + 7);
+		break;
+	case 'monthly':
+		date.setUTCMonth(date.getUTCMonth() + 1);
+		break;
+	case 'yearly':
+		date.setUTCFullYear(date.getUTCFullYear() + 1);
+		break;
+	default:
+		date.setUTCDate(date.getUTCDate() + 1);
+		break;
+	}
+
+	return date;
+};
+
+var filterData = function (data, query) {
+	var result = {}, service, certificate, dateTo, currentDate, mode;
+	service     = certificate;
+	certificate = query.certificate;
+	currentDate = copyDbDate(query.dateFrom);
+	dateTo      = copyDbDate(query.dateTo);
+	modes.some(function (modeItem) {
+		if (modeItem.key === query.mode) {
+			mode = modeItem;
+			return true;
+		}
+	});
+
+	while (currentDate < dateTo) {
+		result[mode.getDisplayedKey(currentDate)] = buildResultRow();
+		incrementDateByTimeUnit(currentDate, mode.key);
```

Comment:
Same problem we have here, `currentDate` needs to be _floored_ up front, otherwise we do not guarantee expected incrementation